### PR TITLE
Add md5 (for HTTP Digest Auth)

### PR DESCRIPTION
**Pull request by cpq**

### CODE REVIEW

File: fossa.c [hunk 1]
```diff
@@ -2484,6 +2484,227 @@ void ns_printf_http_chunk(struct ns_connection *nc, const char *fmt, ...) {
...
+static FILE *open_auth_file(const char *path, int is_directory,
+                            const struct ns_serve_http_opts *opts) {
+  char buf[MAX_PATH_SIZE];
+  const char *p;
+  FILE *fp = NULL;
+
+  if (opts->global_auth_file != NULL) {
+    fp = fopen(opts->global_auth_file, "r");
+  } else if (is_directory && opts->per_directory_auth_file) {
+    snprintf(buf, sizeof(buf), "%s%c%s", path, '/',
```
Comment (mmikulicic) [inline fossa.c]: I guess the idea here is to use a platform specific DIRSEP. Perhaps better do ing it now or at least adding a TODO marker.

File: fossa.c [hunk 2]
```diff
@@ -2484,6 +2484,227 @@ void ns_printf_http_chunk(struct ns_connection *nc, const char *fmt, ...) {
...
+             opts->per_directory_auth_file);
+    fp = fopen(buf, "r");
+  } else if (opts->per_directory_auth_file) {
+    if ((p = strrchr(path, '/')) == NULL) {
+      p = path;
+    }
+    snprintf(buf, sizeof(buf), "%.*s%c%s",
+             (int) (p - path), path, '/', opts->per_directory_auth_file);
```
Comment (mmikulicic) [inline fossa.c]: and here

File: src/http.c [hunk 3]
```diff
@@ -729,6 +729,227 @@ void ns_printf_http_chunk(struct ns_connection *nc, const char *fmt, ...) {
...
+             (int) (p - path), path, '/', opts->per_directory_auth_file);
+    fp = fopen(buf, "r");
+  }
+
+  return fp;
+}
+
+/*
+ * Stringify binary data. Output buffer must be twice as big as input,
+ * because each byte takes 2 bytes in string representation
```
Comment (mmikulicic) [inline src/http.c]: smth like " plus the space for the terminating null byte"

File: fossa.c [hunk 4]
```diff
@@ -2484,6 +2484,227 @@ void ns_printf_http_chunk(struct ns_connection *nc, const char *fmt, ...) {
...
+  }
+  va_end(ap);
+
+  MD5_Final(hash, &ctx);
+  bin2str(buf, hash, sizeof(hash));
+
+  return buf;
+}
+
+static void mkresp(const char *method, size_t method_len,
```
Comment (mmikulicic) [inline fossa.c]: As the static namespace is fossa-wide, `mkresp` is a too common name for what it does.

File: fossa.c [hunk 5]
```diff
@@ -2484,6 +2484,227 @@ void ns_printf_http_chunk(struct ns_connection *nc, const char *fmt, ...) {
...
+}
+
+static int is_authorized(struct http_message *hm, const char *path,
+                         int is_directory, struct ns_serve_http_opts *opts) {
+  FILE *fp;
+  int authorized = 1;
+
+  if (opts->auth_domain == NULL ||
+      (opts->per_directory_auth_file == NULL &&
+       opts->global_auth_file == NULL)) {
```
Comment (mmikulicic) [inline fossa.c]: this empty block is hard to read. Perhaps either a comment or `return 1`

File: fossa.c (hunk 1 above)
Comment (cpq) [inline fossa.c]: Done

File: fossa.c (hunk 2 above)
Comment (cpq) [inline fossa.c]: Done

File: fossa.c (hunk 4 above)
Comment (cpq) [inline fossa.c]: Renamed to `mkmd5resp`

File: fossa.c (hunk 5 above)
Comment (cpq) [inline fossa.c]: Refactored

File: src/http.c (hunk 3 above)
Comment (cpq) [inline src/http.c]: Changed.

**Pull request merged**